src/V3Broken.cpp

```diff
@@ -295,6 +328,62 @@ class BrokenCheckVisitor final : public AstNVisitor {
         UASSERT_OBJ(
             !(v3Global.assertScoped() && m_inScope && nodep->varp() && !nodep->varScopep()), nodep,
             "VarRef missing VarScope pointer");
+        if (m_cfuncp) {
+            // Check if variable is an in-scope local, otherwise mark as suspect
+            if (const AstVar* const varp = nodep->varp()) {
+                if (!isInScopeLocal(varp)) { m_suspectVarRefs.push_back(nodep); }
+            }
+        }
+    }
+    virtual void visit(AstCFunc* nodep) override {
+        UASSERT_OBJ(!m_cfuncp, nodep, "Nested AstCFunc");
+        m_cfuncp = nodep;
+        m_localVars.clear();
+        m_suspectVarRefs.clear();
+        m_localsStack.clear();
+        pushLocalScope();
+
+        processAndIterate(nodep);
+
+        // Check suspect references are all to non-locals
+        for (const AstNodeVarRef* const vrefp : m_suspectVarRefs) {
+            UASSERT_OBJ(m_localVars.count(vrefp->varp()) == 0, vrefp,
+                        "Local variable not in scope where referenced");
+        }
+
+        m_cfuncp = nullptr;
+    }
+    virtual void visit(AstNodeIf* nodep) override {
```

Comment:
Do you need to do anything on While and Jumps?  Probably not but just checking.

Reply:
I don't think so. The if is special because there are 2 (well, strictly 3) scopes within a single statement. We don't have a case statement (at least in CFuncs), and everything else (including AstWhile and AstJumpBlock) is one scope so is covered by the AstNodeStmt case below.